extensions-contrib/spectator-histogram/src/main/java/org/apache/druid/spectator/histogram/SpectatorHistogramAggregatorFactory.java

```diff
@@ -0,0 +1,372 @@
+        cacheTypeId
+    ).appendString(fieldName).build();
+  }
+
+
+  @Override
+  public Aggregator factorize(ColumnSelectorFactory metricFactory)
+  {
+    return new SpectatorHistogramAggregator(metricFactory.makeColumnValueSelector(fieldName));
+  }
+
+  @Override
+  public BufferAggregator factorizeBuffered(ColumnSelectorFactory metricFactory)
+  {
+    return new SpectatorHistogramBufferAggregator(metricFactory.makeColumnValueSelector(fieldName));
+  }
+
+  // This is used when writing metrics to segment files to check whether the column is sorted.
+  // Since there is no sensible way really to compare histograms, compareTo always returns 1.
+  public static final Comparator<SpectatorHistogram> COMPARATOR = (o, o1) -> {
+    if (o == null && o1 == null) {
+      return 0;
+    } else if (o != null && o1 == null) {
+      return -1;
+    } else if (o == null) {
+      return 1;
+    }
+    return Integer.compare(o.hashCode(), o1.hashCode());
+  };
+
+  @Override
+  public Comparator getComparator()
+  {
+    return COMPARATOR;
+  }
+
+  @Override
+  public Object combine(@Nullable Object lhs, @Nullable Object rhs)
+  {
+    if (lhs == null) {
+      return rhs;
+    }
+    if (rhs == null) {
+      return lhs;
+    }
+    SpectatorHistogram lhsHisto = (SpectatorHistogram) lhs;
+    SpectatorHistogram rhsHisto = (SpectatorHistogram) rhs;
+    lhsHisto.merge(rhsHisto);
+    return lhsHisto;
+  }
+
+  @Override
+  public AggregatorFactory getCombiningFactory()
+  {
+    return new SpectatorHistogramAggregatorFactory(name, name);
+  }
+
+  @Override
+  public AggregatorFactory getMergingFactory(AggregatorFactory other) throws AggregatorFactoryNotMergeableException
+  {
+    if (other.getName().equals(this.getName()) && this.getClass() == other.getClass()) {
+      return getCombiningFactory();
+    } else {
+      throw new AggregatorFactoryNotMergeableException(this, other);
+    }
+  }
+
+  @Override
+  public List<AggregatorFactory> getRequiredColumns()
+  {
+    return Collections.singletonList(
+        new SpectatorHistogramAggregatorFactory(
+            fieldName,
+            fieldName
+        )
+    );
+  }
+
+  @Override
+  public Object deserialize(Object serializedHistogram)
+  {
+    return SpectatorHistogram.deserialize(serializedHistogram);
+  }
+
+  @Nullable
+  @Override
+  public Object finalizeComputation(@Nullable Object object)
+  {
+    return object;
+  }
+
+  @Override
+  @JsonProperty
+  public String getName()
+  {
+    return name;
+  }
+
+  @JsonProperty
+  public String getFieldName()
+  {
+    return fieldName;
+  }
+
+  @Override
+  public List<String> requiredFields()
+  {
+    return Collections.singletonList(fieldName);
+  }
+
+  @Override
+  public String getComplexTypeName()
+  {
+    return TYPE_NAME;
+  }
+
+  @Override
+  public ValueType getType()
+  {
+    return ValueType.COMPLEX;
+  }
+
+  @Override
+  public ValueType getFinalizedType()
+  {
+    return ValueType.COMPLEX;
+  }
```

Comment:
these methods are deprecated and will be removed at some point, please implement `getIntermediateType` and `getResultType` instead.